jupyterhubutils/lsstmgr/namespacemanager.py

```diff
@@ -0,0 +1,298 @@
+'''Class to provide namespace manipulation.
+'''
+
+import time
+from kubernetes.client.rest import ApiException
+from kubernetes import client
+
+from ..utils import (get_execution_namespace, make_logger)
+
+
+class LSSTNamespaceManager(object):
+    '''Class to provide namespace manipulation.
+    '''
+    namespace = None
+    service_account = None
+
+    def __init__(self, *args, **kwargs):
+        self.log = make_logger()
+        self.log.debug("Creating LSSTNamespaceManager")
+        self.parent = kwargs.pop('parent')
+
+    def update_namespace_name(self):
+        '''Build namespace name from user and execution namespace.
+        '''
+        execution_namespace = get_execution_namespace()
+        self.log.debug("Execution namespace: '{}'".format(execution_namespace))
+        user = self.parent.user
+        self.log.debug("User: '{}'".format(user))
+        username = user.escaped_name
+        if execution_namespace and username:
+            self.namespace = "{}-{}".format(execution_namespace,
+                                            username)
+        else:
+            df_msg = "Using 'default' namespace."
+            self.log.warning(df_msg)
+            self.namespace = "default"
+
+    def ensure_namespace(self):
+        '''Here we make sure that the namespace exists, creating it if
+        it does not.  That requires a ClusterRole that can list and create
+        namespaces.
+
+        If we have shadow PVs, we clone the (static) NFS PVs and then
+        attach namespaced PVCs to them.  Thus the role needs to be
+        able to list and create PVs and PVCs.
+
+        If we create the namespace, we also create (if needed) a ServiceAccount
+        within it to allow the user pod to spawn dask and workflow pods.
+
+        '''
+        self.update_namespace_name()
```

Comment:
make this a private method _update_namespace_name() and only call it here

Reply:
Spawner will set it when it's time